File: Data Structures & Algorithms/binary-tree-right-side-view/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> res;
        if (!root) {
            return res;
        }
        queue<pair<TreeNode*,int>> q;
        int curLevel = 0;
        q.push({root, 1});

        while (!q.empty()) {
            TreeNode* node = q.front().first;
            int l = q.front().second;
            
            q.pop();

            if (l != curLevel) {
                res.push_back(node->val);
                curLevel++;
            }
            if (node->right) {
                q.push({node->right, l + 1});
            }
            if (node->left) {
                q.push({node->left, l + 1});
            }
        }
        return res;
    }
};

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        stack = [root]
        while stack:
            node = stack.pop()
            if node.val == subRoot.val:
                subStack = [(node, subRoot)]
                isEqual = True
                while subStack:
                    aNode, bNode = subStack.pop()
                    if not aNode and not bNode:
                        continue
                    if not aNode or not bNode or aNode.val != bNode.val:
                        isEqual = False
                        break
                    subStack.append((aNode.left, bNode.left))
                    subStack.append((aNode.right, bNode.right))
                if isEqual:
                    return True
            if node.left:
                stack.append(node.left)
            if node.right:
                stack.append(node.right)
        return False